### PR TITLE
Azure Connector implementation

### LINKED ISSUE #1069: Data Connector: Azure Data Lake Storage Gen2 (ADLSv2)
With support for [Hierachical namespace](https://learn.microsoft.com/en-us/azure/storage/blobs/data-lake-storage-namespace?source=recommendations)

Example:

```yaml
datasets:
  - from: adls2:/container/path
     name: datalake_storage_path
```

Supported formats:

- JSON
- CSV
- Parquet

### LINKED ISSUE #2917: Enhancement: Azure data connector
REMEMBER, BE **SMART**!

S: Specific
M: Measurable
A: Achievable
R: Relevant
T: Time-Bound

## Goal-State/What/Result

This connector would enable users to use all of the supported Azure filesystems (ADLS, Azure Blobs) as a source system for all of their data.

## Why/Purpose

Enabling a wider support for various cloud architectures. It's also common for data to live in multiple locations and across cloud providers for redundancy

## By When

A target date for done-done completion of the entire enhancement.

**Issue/Spec written and reviewed:** <Target Date>
**Done-Done:** Friday, October 4th

## Done-Done

- [ ] [First Principles Driven](https://github.com/spiceai/spiceai/blob/trunk/docs/PRINCIPLES.md)
- [ ] The Algorithm
- [x] PM/Design Review
- [x] UX Review
- [ ] Threat Model / Security Review
- [ ] Tests
- [ ] Documentation
- [ ] Samples/Quickstarts/Tutorials
- [ ] Release Notes

## The Algorithm

- [ ] Every requirement questioned?
- [ ] Delete (Scope) any part you can.
- [ ] Simplify.
- [ ] Break down into smaller iterations/milestones.
- [ ] Opportunities for automation.

## Spec

The Azure connector (now known as the ABFS connector, will have the following Spicepod spec:

```yaml
datasets:
  - from: abfs://foocontainer/taxi_sample.csv
    name: azure_test
    params:
      abfs_account: spicescottadls
      abfs_access_key: abc123
```

The URL follows the format of `abfs://container_name/file_path`

All parameters supported by `object_store` will be supported in the `params` for this connector, prefixed with `abfs`:

- account
- container_name
- access_key
- bearer_token
- client_id
- client_secret
- tenant_id
- sas_string
- endpoint
- use_emulator
- use_fabric_endpoint
- allow_http
- authority_host
- max_retries
- retry_timeout
- backoff_initial_duration
- backoff_max_duration
- backoff_base
- proxy_url
- proxy_ca_certificate
- proxy_excludes
- msi_endpoint
- federated_token_file
- use_cli
- skip_signature
- disable_tagging

## How/MAP

A clear and concise plan of how this should be implemented.

### Implementation (issues)

- 
-

### PR DESCRIPTION
**Pull request by eadgbear**
## 🗣 Description

This is a wrapper around the existing `object_store` support for Azure. 

Test scenarios:

- [ X ] Local emulator
- [ X ] Remote public account/container

## 🔨 Related Issues

- Closes #2917
- Closes #1069  

## 📄 Documentation Requirements

- https://github.com/spiceai/docs/issues/432
- I don't think a separate quickstart is necessary, but a sample showing other Azure integrations would be useful

### CODE REVIEW

File: crates/runtime/src/dataconnector/azure.rs [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Unable to parse URL {url}: {source}"))]
+    UnableToParseURL {
+        url: String,
+        source: url::ParseError,
+    },
+    #[snafu(display(
+        "Only one of access_key, bearer token or client credentials must be provided"
```
Comment (lukekim) [inline crates/runtime/src/dataconnector/azure.rs]:
```suggestion
        "Provide only one of the following: access key, bearer token, or client credentials."
```

File: crates/runtime/src/dataconnector/azure.rs [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+
+    #[must_use]
+    pub fn new_arc(prefix: &'static str) -> Arc<dyn DataConnectorFactory> {
+        Arc::new(Self { prefix }) as Arc<dyn DataConnectorFactory>
+    }
+}
+
+const PARAMETERS: &[ParameterSpec] = &[
+    ParameterSpec::connector("account")
+        .description("The Azure Storage account name.")
```
Comment (lukekim) [inline crates/runtime/src/dataconnector/azure.rs]: Remove `The ` from all these.

File: crates/runtime/src/dataconnector/azure.rs [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    ParameterSpec::runtime("csv_has_header")
+        .description("Set true to indicate that the first line is a header."),
+    ParameterSpec::runtime("csv_quote").description("The quote character in a row."),
+    ParameterSpec::runtime("csv_escape").description("The escape character in a row."),
+    ParameterSpec::runtime("csv_schema_infer_max_records")
+        .description("Set a limit in terms of records to scan to infer the schema."),
+    ParameterSpec::runtime("csv_delimiter")
+        .description("The character separating values within a row."),
+    ParameterSpec::runtime("file_compression_type")
+        .description("The type of compression used on the file. Supported types are: GZIP, BZIP2, XZ, ZSTD, UNCOMPRESSED"),
```
Comment (lukekim) [inline crates/runtime/src/dataconnector/azure.rs]: Lowercase values.

File: crates/runtime/src/dataconnector/azure.rs [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+        })
+    }
+
+    fn prefix(&self) -> &'static str {
+        &self.prefix
+    }
+
+    // We need a unified parameter prefix for all Azure connectors
+    fn parameter_prefix(&self) -> &'static str {
+        "azure"
```
Comment (lukekim) [inline crates/runtime/src/dataconnector/azure.rs]: I don't think we should have this. We should have a common data component, and multiple connectors with specific prefixes. In this case, I am thinking `azstorage` makes the most sense?

File: crates/runtime/tests/integration.rs [hunk 5]
```diff
@@ -27,6 +27,7 @@ use tracing::subscriber::DefaultGuard;
 use tracing_subscriber::EnvFilter;
 
 mod acceleration;
+mod azure;
```
Comment (lukekim) [inline crates/runtime/tests/integration.rs]: azstorage

File: crates/runtime/src/dataconnector/azure.rs [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+    ParameterSpec::runtime("csv_escape").description("The escape character in a row."),
+    ParameterSpec::runtime("csv_schema_infer_max_records")
+        .description("Set a limit in terms of records to scan to infer the schema."),
+    ParameterSpec::runtime("csv_delimiter")
+        .description("The character separating values within a row."),
+    ParameterSpec::runtime("file_compression_type")
+        .description("The type of compression used on the file. Supported types are: GZIP, BZIP2, XZ, ZSTD, UNCOMPRESSED"),
+];
+
+pub const PREFIXES: [&'static str; 6] = ["azure", "abfs", "abfss", "adl", "adls", "az"];
```
Comment (lukekim) [inline crates/runtime/src/dataconnector/azure.rs]: You shouldn't need to do this if the main prefix is `azstorage`. These should be proper schema prefixes E.g. `abfs://` and on false, an error should be thrown, not just a continuation.

**Review by lukekim — DISMISSED** [review]
Need to resolve the spicepod DX.

File: crates/runtime/src/object_store_registry.rs [hunk 7]
```diff
@@ -203,13 +207,201 @@ impl SpiceObjectStoreRegistry {
...
+            if !as_bool {
+                builder = builder.with_url(url.to_string());
+            }
+        } else {
+            builder = builder.with_url(url.to_string());
+        }
+
+        builder = builder.with_url(url.to_string());
+
+        if let Some(account) = params.get("account") {
```
Comment (sgrebnov) [inline crates/runtime/src/object_store_registry.rs:247]: Nit: If there are some required combination of parameters, like `skip_signature` if there are no credentials specified, we should help user to understand the problem (user friendly error). Will MicrosoftAzureBuilder handle this nicely?

File: crates/runtime/tests/azure/mod.rs [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+    blob_client
+        .put_block_blob(sample_file)
+        .content_type("text/csv")
+        .await?;
+    tracing::trace!("Sample file uploaded");
+    Ok(())
+}
+
+#[tokio::test]
+async fn test_azure_connector() -> Result<(), anyhow::Error> {
```
Comment (sgrebnov) [inline crates/runtime/tests/azure/mod.rs]: Run tests locally and everything worked well 👍

File: crates/runtime/tests/azure/mod.rs [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+    azurite_container.stop().await?;
+    azurite_container.remove().await?;
+    res.map_err(|e| anyhow::anyhow!(e))
+}
+
+fn make_test_query(table_name: &str) -> String {
+    format!("SELECT DISTINCT(\"VendorID\") FROM {table_name} ORDER BY \"VendorID\" DESC")
+}
+
+async fn run_test() -> Result<(), anyhow::Error> {
```
Comment (sgrebnov) [inline crates/runtime/tests/azure/mod.rs]: Probably `move upload_sample_file().await?;` to preparation step and rename to `test_spice_with_azstorage_datasets`

Comment (eadgbear) [reply crates/runtime/tests/azure/mod.rs]: Refactored!

File: crates/runtime/tests/azure/mod.rs (hunk 8 above)
Comment (eadgbear) [reply crates/runtime/tests/azure/mod.rs]: Glad to hear it!

File: crates/runtime/src/object_store_registry.rs (hunk 7 above)
Comment (eadgbear) [reply crates/runtime/src/object_store_registry.rs:247]: I've added a condition to the actual connector itself to test for this connection, we'll throw an error if there's a incorrect combination of credentials.

File: crates/runtime/src/dataconnector/azure.rs (hunk 3 above)
Comment (eadgbear) [reply crates/runtime/src/dataconnector/azure.rs]: Done

File: crates/runtime/src/dataconnector/azure.rs (hunk 6 above)
Comment (eadgbear) [reply crates/runtime/src/dataconnector/azure.rs]: I went digging through the Microsoft docs and it turns out that the preferred prefix is `abfss`, so I've renamed the connector to that.

**Review by phillipleblanc — DISMISSED** [review]
This looks great to me, thanks!

**Review by sgrebnov — DISMISSED** [review]

**Review by sgrebnov — APPROVED** [review]

**Pull request merged**